presto-main/src/main/java/com/facebook/presto/cost/EnforceSingleRowStatsRule.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.Map;
+import java.util.Optional;
+
+public class EnforceSingleRowStatsRule
+        implements ComposableStatsCalculator.Rule
+{
+    private static final Pattern<EnforceSingleRowNode> PATTERN = Pattern.typeOf(EnforceSingleRowNode.class);
+
+    @Override
+    public Pattern<EnforceSingleRowNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> calculate(PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        // TODO retain useful information from source' stats
```

Comment:
This requires "normalizers" to avoid copying logic -- "normalizers" are a way to derived certain properties when something is already known. E.g. when we know row count to be 1, we can cap all NDVs to be at most 1.

"normalizers" is something that needs rework and i am working on this now. Won't be part of this PR. Also, specifically for `EnforceSingleRowNode` I think this is not a game changer.